modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java

```diff
@@ -0,0 +1,709 @@
+import org.mule.runtime.module.extension.api.loader.java.type.WithAnnotations;
+import org.mule.sdk.api.annotation.DoNotEnforceMinMuleVersion;
+import org.mule.sdk.api.annotation.Extension;
+import org.mule.sdk.api.annotation.MinMuleVersion;
+
+import javax.inject.Inject;
+import java.lang.annotation.Annotation;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+import static org.mule.runtime.api.meta.MuleVersion.FIRST_MULE_VERSION;
+
+/**
+ * Utils class to create {@link MinMuleVersionResult}s from {@link Type}s.
+ *
+ * @since 4.5
+ */
+public final class MinMuleVersionUtils {
+
+  private static final MuleVersion SDK_EXTENSION_ANNOTATION_MIN_MULE_VERSION =
+      new MuleVersion(Extension.class.getAnnotation(MinMuleVersion.class).value());
+  private static final String REASON_ANNOTATED = "has min mule version %s because it is annotated with %s.";
+  private static final String REASON_SUPER_CLASS = "has min mule version %s because of its super class %s.";
+  private static final String REASON_FIELD = "has min mule version %s because of its field %s.";
+  private static final String REASON_INTERFACE = "has min mule version %s because it implements interface %s.";
+  private static final String REASON_PARAMETER = "has min mule version %s because of its parameter %s.";
+  private static final String REASON_TYPE = "has min mule version %s because it is of type %s.";
+  private static final String REASON_METHOD = "has min mule version %s because of its method %s.";
+  private static final String REASON_CLASS_LEVEL_MMV =
+      " has min mule version %s because it is the one set at the class level through the @MinMuleVersion annotation.";
+
+  private MinMuleVersionUtils() {}
+
+  /**
+   * Calculates the minimum {@link MuleVersion} for a given {@link ExtensionElement} by looking at the class annotation
+   * {@link MinMuleVersion} (if present) and at the new SDK API components it interacts with through its annotations, fields and
+   * methods.
+   *
+   * @param extension the extension to calculate its min mule version
+   * @return the minimum mule version of the given extension
+   */
+  public static MinMuleVersionResult getExtensionResult(Type extension) {
+    if (belongsToJavaPackages(extension.getTypeName())) {
+      return createResultWithDefaultMMV("Java type", extension.getTypeName());
+    }
+    MinMuleVersionResult extensionResult;
+    if (extension.isAnnotatedWith(Extension.class)) {
+      extensionResult = new MinMuleVersionResult(extension.getName(), SDK_EXTENSION_ANNOTATION_MIN_MULE_VERSION,
+                                                 format("Extension %s has min mule version %s because it if annotated with the new sdk api @Extension.",
```

Comment:
do not user String.format() here. THat method is VERY slow and will affect Mule startup time.